File: Farkle/Models/GameEngine.swift
import Foundation
import SwiftData

@MainActor
struct GameEngine {
    let game: Game
    let context: ModelContext

    func bank(hotDice: Bool = false) {
        guard let player = game.activePlayer else { return }
        let amount = game.pendingTurnScore
        guard amount > 0 else { return }
        if let mustOpen = game.rules.mustOpenWith, player.bankedScore == 0, amount < mustOpen { return }

        let entry = ActionLogEntry(playerID: player.id,
                                   kind: .bank,
                                   amount: amount,
                                   orderIndex: game.actions.count,
                                   roundNumber: game.currentRound,
                                   hotDice: hotDice,
                                   pendingTurnAtAction: amount)
        game.actions.append(entry)
        player.bankedScore += amount
        if hotDice { player.hotDiceCount += 1 }
        game.pendingTurnScore = 0
        game.pendingRollCount = 0

        if game.finalRoundTriggeredByPlayerID == nil, player.bankedScore >= game.targetScore {
            game.finalRoundTriggeredByPlayerID = player.id
            game.finalRoundTurnsRemaining = max(0, game.orderedPlayers.count - 1)
            game.finalRoundAnnouncementShown = false
            let trigger = ActionLogEntry(playerID: player.id,
                                         kind: .startFinalRound,
                                         amount: 0,
                                         orderIndex: game.actions.count,
                                         roundNumber: game.currentRound,
                                         pendingTurnAtAction: 0)
            game.actions.append(trigger)
        } else if game.finalRoundTriggeredByPlayerID != nil {
            game.finalRoundTurnsRemaining = max(0, game.finalRoundTurnsRemaining - 1)
        }

        advancePlayerOrEndGame()
        save()
    }

    func bust() {
        guard let player = game.activePlayer else { return }
        let snapshot = game.pendingTurnScore
        let entry = ActionLogEntry(playerID: player.id,
                                   kind: .bust,
                                   amount: 0,
                                   orderIndex: game.actions.count,
                                   roundNumber: game.currentRound,
                                   pendingTurnAtAction: snapshot)
        game.actions.append(entry)
        player.farkleCount += 1
        game.pendingTurnScore = 0
        game.pendingRollCount = 0

        if game.finalRoundTriggeredByPlayerID != nil {
            game.finalRoundTurnsRemaining = max(0, game.finalRoundTurnsRemaining - 1)
        }
        advancePlayerOrEndGame()
        save()
    }

    func addToPending(_ amount: Int) {
        guard amount > 0 else { return }
        game.pendingTurnScore += amount
        game.pendingRollCount += 1
        save()
    }

    func setPending(_ amount: Int) {
        game.pendingTurnScore = max(0, amount)
        save()
    }

    func clearPending() {
        game.pendingTurnScore = 0
        game.pendingRollCount = 0
        save()
    }

    /// Undo the most recent action. Returns true if something was undone.
    @discardableResult
    func undoLast() -> Bool {
        guard let last = game.orderedActions.last else { return false }
        return undo(through: last.orderIndex)
    }

    /// Undo the specific action by id, plus everything after it.
    @discardableResult
    func undo(actionID: UUID) -> Bool {
        guard let target = game.actions.first(where: { $0.id == actionID }) else { return false }
        return undo(through: target.orderIndex)
    }

    private func undo(through orderIndex: Int) -> Bool {
        let toRemove = game.actions.filter { $0.orderIndex >= orderIndex }
        guard !toRemove.isEmpty else { return false }
        let removeIDs = Set(toRemove.map(\.id))
        game.actions.removeAll { removeIDs.contains($0.id) }
        for a in toRemove { context.delete(a) }
        rebuildDerivedState()
        save()
        return true
    }

    private func rebuildDerivedState() {
        for p in game.players {
            p.bankedScore = 0
            p.hotDiceCount = 0
            p.farkleCount = 0
        }
        game.activePlayerIndex = 0
        game.pendingTurnScore = 0
        game.pendingRollCount = 0
        game.finalRoundTriggeredByPlayerID = nil
        game.finalRoundTurnsRemaining = 0
        game.finalRoundAnnouncementShown = true
        game.winnerPlayerID = nil
        game.endedAt = nil

        let ordered = game.orderedActions
        let players = game.orderedPlayers
        var activeIdx = 0

        for entry in ordered {
            guard let player = players.first(where: { $0.id == entry.playerID }) else { continue }
            switch entry.kind {
            case .bank:
                player.bankedScore += entry.amount
                if entry.hotDice { player.hotDiceCount += 1 }
                if game.finalRoundTriggeredByPlayerID != nil {
                    game.finalRoundTurnsRemaining = max(0, game.finalRoundTurnsRemaining - 1)
                }
                activeIdx = (activeIdx + 1) % max(1, players.count)
            case .bust:
                player.farkleCount += 1
                if game.finalRoundTriggeredByPlayerID != nil {
                    game.finalRoundTurnsRemaining = max(0, game.finalRoundTurnsRemaining - 1)
                }
                activeIdx = (activeIdx + 1) % max(1, players.count)
            case .startFinalRound:
                game.finalRoundTriggeredByPlayerID = entry.playerID
                game.finalRoundTurnsRemaining = max(0, players.count - 1)
                game.finalRoundAnnouncementShown = false
            case .endGame:
                game.endedAt = entry.timestamp
                game.winnerPlayerID = entry.playerID
            }
        }
        game.activePlayerIndex = activeIdx
    }

    private func advancePlayerOrEndGame() {
        let players = game.orderedPlayers
        guard !players.isEmpty else { return }

        if game.finalRoundTriggeredByPlayerID != nil, game.finalRoundTurnsRemaining <= 0 {
            // game over
            let leader = players.max(by: { $0.bankedScore < $1.bankedScore })
            game.winnerPlayerID = leader?.id
            game.endedAt = Date()
            let endEntry = ActionLogEntry(playerID: leader?.id ?? UUID(),
                                          kind: .endGame,
                                          amount: leader?.bankedScore ?? 0,
                                          orderIndex: game.actions.count,
                                          roundNumber: game.currentRound)
            game.actions.append(endEntry)
        } else {
            game.activePlayerIndex = (game.activePlayerIndex + 1) % players.count
        }
    }

    /// Called after the user dismisses the final-round announcement screen.
    func markFinalRoundAnnouncementShown() {
        game.finalRoundAnnouncementShown = true
        save()
    }

    /// Replace the amount on a past bank action. The action keeps its slot in the log,
    /// player order isn't affected; derived totals are rebuilt by replay. No-op for
    /// non-bank actions and for non-positive amounts.
    @discardableResult
    func setActionAmount(actionID: UUID, newAmount: Int) -> Bool {
        guard let target = game.actions.first(where: { $0.id == actionID }),
              target.kind == .bank,
              newAmount > 0 else { return false }
        target.amount = newAmount
        target.pendingTurnAtAction = newAmount
        rebuildDerivedState()
        save()
        return true
    }

    func renamePlayer(_ player: Player, to newName: String) {
        let trimmed = newName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        player.name = String(trimmed.prefix(20))
        save()
    }

    private func save() {
        try? context.save()
    }
}

File: android/app/src/main/java/com/feltandbone/farkle/engine/GameEngine.kt
package com.feltandbone.farkle.engine

import com.feltandbone.farkle.model.ActionKind
import com.feltandbone.farkle.model.ActionLogEntry
import com.feltandbone.farkle.model.Game
import com.feltandbone.farkle.model.Player
import java.util.UUID
import kotlin.math.max

/**
 * Pure game-state transitions. Every function takes a [Game] and returns a new [Game];
 * derived totals are always rebuildable by replaying the action log. Parity with the
 * iOS `GameEngine`.
 */
object GameEngine {

    private fun now() = System.currentTimeMillis()

    fun bank(game: Game, hotDice: Boolean = false): Game {
        val player = game.activePlayer ?: return game
        val amount = game.pendingTurnScore
        if (amount <= 0) return game
        val mustOpen = game.rules.mustOpenWith
        if (mustOpen != null && player.bankedScore == 0 && amount < mustOpen) return game

        val roundForBank = game.currentRound
        val entry = ActionLogEntry(
            playerId = player.id,
            kind = ActionKind.BANK,
            amount = amount,
            timestamp = now(),
            orderIndex = game.actions.size,
            roundNumber = roundForBank,
            hotDice = hotDice,
            pendingTurnAtAction = amount,
        )
        val players = game.players.map {
            if (it.id == player.id) {
                it.copy(
                    bankedScore = it.bankedScore + amount,
                    hotDiceCount = it.hotDiceCount + if (hotDice) 1 else 0,
                )
            } else it
        }
        var g = game.copy(
            players = players,
            actions = game.actions + entry,
            pendingTurnScore = 0,
            pendingRollCount = 0,
        )

        val updated = players.first { it.id == player.id }
        if (g.finalRoundTriggeredByPlayerId == null && updated.bankedScore >= g.targetScore) {
            val trigger = ActionLogEntry(
                playerId = player.id,
                kind = ActionKind.START_FINAL_ROUND,
                amount = 0,
                timestamp = now(),
                orderIndex = g.actions.size,
                roundNumber = g.currentRound,
                pendingTurnAtAction = 0,
            )
            g = g.copy(
                finalRoundTriggeredByPlayerId = player.id,
                finalRoundTurnsRemaining = max(0, g.orderedPlayers.size - 1),
                finalRoundAnnouncementShown = false,
                actions = g.actions + trigger,
            )
        } else if (g.finalRoundTriggeredByPlayerId != null) {
            g = g.copy(finalRoundTurnsRemaining = max(0, g.finalRoundTurnsRemaining - 1))
        }

        return advancePlayerOrEndGame(g)
    }

    fun bust(game: Game): Game {
        val player = game.activePlayer ?: return game
        val snapshot = game.pendingTurnScore
        val entry = ActionLogEntry(
            playerId = player.id,
            kind = ActionKind.BUST,
            amount = 0,
            timestamp = now(),
            orderIndex = game.actions.size,
            roundNumber = game.currentRound,
            pendingTurnAtAction = snapshot,
        )
        val players = game.players.map {
            if (it.id == player.id) it.copy(farkleCount = it.farkleCount + 1) else it
        }
        var g = game.copy(
            players = players,
            actions = game.actions + entry,
            pendingTurnScore = 0,
            pendingRollCount = 0,
        )
        if (g.finalRoundTriggeredByPlayerId != null) {
            g = g.copy(finalRoundTurnsRemaining = max(0, g.finalRoundTurnsRemaining - 1))
        }
        return advancePlayerOrEndGame(g)
    }

    fun addToPending(game: Game, amount: Int): Game {
        if (amount <= 0) return game
        return game.copy(
            pendingTurnScore = game.pendingTurnScore + amount,
            pendingRollCount = game.pendingRollCount + 1,
        )
    }

    fun setPending(game: Game, amount: Int): Game = game.copy(pendingTurnScore = max(0, amount))

    fun clearPending(game: Game): Game = game.copy(pendingTurnScore = 0, pendingRollCount = 0)

    fun undoLast(game: Game): Game {
        val last = game.orderedActions.lastOrNull() ?: return game
        return undoThrough(game, last.orderIndex)
    }

    fun undo(game: Game, actionId: String): Game {
        val target = game.actions.firstOrNull { it.id == actionId } ?: return game
        return undoThrough(game, target.orderIndex)
    }

    private fun undoThrough(game: Game, orderIndex: Int): Game {
        val remaining = game.actions.filter { it.orderIndex < orderIndex }
        if (remaining.size == game.actions.size) return game
        return rebuildDerivedState(game.copy(actions = remaining))
    }

    /** Replace the amount on a past bank action. The action keeps its slot; totals are replayed. */
    fun setActionAmount(game: Game, actionId: String, newAmount: Int): Game {
        val target = game.actions.firstOrNull { it.id == actionId } ?: return game
        if (target.kind != ActionKind.BANK || newAmount <= 0) return game
        val newActions = game.actions.map {
            if (it.id == actionId) it.copy(amount = newAmount, pendingTurnAtAction = newAmount) else it
        }
        return rebuildDerivedState(game.copy(actions = newActions))
    }

    /** Re-assign orderIndex so players sort in the given order; the same physical player stays active. */
    fun reorderPlayers(game: Game, ids: List<String>): Game {
        val activeId = game.activePlayer?.id
        val players = game.players.map { p ->
            val idx = ids.indexOf(p.id)
            if (idx >= 0) p.copy(orderIndex = idx) else p
        }
        var g = game.copy(players = players)
        if (activeId != null) {
            val newIdx = g.orderedPlayers.indexOfFirst { it.id == activeId }
            if (newIdx >= 0) g = g.copy(activePlayerIndex = newIdx)
        }
        return g
    }

    /** Append a new player. Only allowed during the first round (see [Game.canAddPlayer]). */
    fun addPlayer(game: Game, name: String): Game {
        if (!game.canAddPlayer) return game
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return game
        val activeId = game.activePlayer?.id
        val nextIndex = game.orderedPlayers.size
        val player = Player(
            name = trimmed.take(20),
            avatarIndex = nextIndex,
            orderIndex = nextIndex,
        )
        var g = game.copy(players = game.players + player)
        if (activeId != null) {
            val newIdx = g.orderedPlayers.indexOfFirst { it.id == activeId }
            if (newIdx >= 0) g = g.copy(activePlayerIndex = newIdx)
        }
        return g
    }

    fun renamePlayer(game: Game, playerId: String, newName: String): Game {
        val trimmed = newName.trim()
        if (trimmed.isEmpty()) return game
        val players = game.players.map {
            if (it.id == playerId) it.copy(name = trimmed.take(20)) else it
        }
        return game.copy(players = players)
    }

    fun markFinalRoundAnnouncementShown(game: Game): Game =
        game.copy(finalRoundAnnouncementShown = true)

    private fun advancePlayerOrEndGame(game: Game): Game {
        val players = game.orderedPlayers
        if (players.isEmpty()) return game
        return if (game.finalRoundTriggeredByPlayerId != null && game.finalRoundTurnsRemaining <= 0) {
            val leader = players.maxByOrNull { it.bankedScore }
            val endEntry = ActionLogEntry(
                playerId = leader?.id ?: UUID.randomUUID().toString(),
                kind = ActionKind.END_GAME,
                amount = leader?.bankedScore ?: 0,
                timestamp = now(),
                orderIndex = game.actions.size,
                roundNumber = game.currentRound,
            )
            game.copy(
                winnerPlayerId = leader?.id,
                endedAt = now(),
                actions = game.actions + endEntry,
            )
        } else {
            game.copy(activePlayerIndex = (game.activePlayerIndex + 1) % players.size)
        }
    }

    private fun rebuildDerivedState(game: Game): Game {
        var players = game.players.map { it.copy(bankedScore = 0, hotDiceCount = 0, farkleCount = 0) }
        val playerCount = game.orderedPlayers.size
        var activeIdx = 0
        var finalTrigger: String? = null
        var finalRemaining = 0
        var finalAnnouncementShown = true
        var endedAt: Long? = null
        var winnerId: String? = null

        for (entry in game.orderedActions) {
            val idx = players.indexOfFirst { it.id == entry.playerId }
            when (entry.kind) {
                ActionKind.BANK -> {
                    if (idx >= 0) {
                        players = players.toMutableList().also {
                            it[idx] = it[idx].copy(
                                bankedScore = it[idx].bankedScore + entry.amount,
                                hotDiceCount = it[idx].hotDiceCount + if (entry.hotDice) 1 else 0,
                            )
                        }
                    }
                    if (finalTrigger != null) finalRemaining = max(0, finalRemaining - 1)
                    activeIdx = (activeIdx + 1) % max(1, playerCount)
                }
                ActionKind.BUST -> {
                    if (idx >= 0) {
                        players = players.toMutableList().also {
                            it[idx] = it[idx].copy(farkleCount = it[idx].farkleCount + 1)
                        }
                    }
                    if (finalTrigger != null) finalRemaining = max(0, finalRemaining - 1)
                    activeIdx = (activeIdx + 1) % max(1, playerCount)
                }
                ActionKind.START_FINAL_ROUND -> {
                    finalTrigger = entry.playerId
                    finalRemaining = max(0, playerCount - 1)
                    finalAnnouncementShown = false
                }
                ActionKind.END_GAME -> {
                    endedAt = entry.timestamp
                    winnerId = entry.playerId
                }
            }
        }

        return game.copy(
            players = players,
            activePlayerIndex = activeIdx,
            pendingTurnScore = 0,
            pendingRollCount = 0,
            finalRoundTriggeredByPlayerId = finalTrigger,
            finalRoundTurnsRemaining = finalRemaining,
            finalRoundAnnouncementShown = finalAnnouncementShown,
            winnerPlayerId = winnerId,
            endedAt = endedAt,
        )
    }
}
